File: src/interface/src/Lagoon/Interface/Source.hs
-- Copyright 2020 Pfizer Inc.

-- Licensed under the Apache License, Version 2.0 (the "License");
-- you may not use this file except in compliance with the License.
-- You may obtain a copy of the License at

--     https://www.apache.org/licenses/LICENSE-2.0

-- Unless required by applicable law or agreed to in writing, software
-- distributed under the License is distributed on an "AS IS" BASIS,
-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-- See the License for the specific language governing permissions and
-- limitations under the License.
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE OverloadedStrings #-}
module Lagoon.Interface.Source (
    SourceIx
  , Version(..)
  , SourceName
  , SourceNameIx(..)
  , Source(..)
  , Description
  , TagName
  ) where

import Data.Aeson
import GHC.Generics (Generic)
import Text.Show.Pretty
import Web.HttpApiData

import Lagoon.Interface.DB
import Lagoon.Interface.Pretty

-- | Source ID
--
-- TODO: Make this a newtype.
type SourceIx = Ix

-- | Source version
newtype Version = Version Int
  deriving (Show, PrettyVal, Pretty, FromJSON, ToJSON)

-- | Source name (identifying a particular dataset)
type SourceName = String

-- | Source description
type Description = String

-- |  Tags
type TagName = String

-- | Source name ID (in other words, ID of a version of a dataset)
newtype SourceNameIx = SourceNameIx Ix
  deriving (Show, PrettyVal, Pretty, FromJSON, ToJSON, FromHttpApiData, ToHttpApiData)

-- | Source name along with its ID
data Source = Source {
      sourceNameIx :: SourceNameIx
    , sourceName   :: SourceName
    }
  deriving (Show, Generic)

instance Pretty Source where
  pretty = pretty . sourceName

instance PrettyVal Source

{-------------------------------------------------------------------------------
  Serialization
-------------------------------------------------------------------------------}

instance ToJSON Source where
  toJSON Source{..} = object [
      "name" .= sourceName
    , "ix"   .= sourceNameIx
    ]

instance FromJSON Source where
  parseJSON = withObject "Source" $ \obj -> do
    sourceName   <- obj .: "name"
    sourceNameIx <- obj .: "ix"
    return Source{..}

instance ToHttpApiData Version where
  toQueryParam (Version i) = toQueryParam i

instance FromHttpApiData Version where
  parseQueryParam = fmap Version . parseQueryParam

File: clients/PyLagoon/PyLagoon/source.py
# Copyright 2020 Pfizer Inc.

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#     https://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
import pandas as pd
import numpy as np

LAGOON_TYPES = {
    "BOOLEAN": bool,
    "INTEGER": int,
    "BIGINT": pd.Int64Dtype,
    "DOUBLE PRECISION": np.float64,
    "TEXT": str,
    "ARR": object,
    "DOCUMENT": object,
    "JSON": object,
}
UNKNOWN_COLUMN_TYPE = object


class Source:
    """A wrapper for the json description of sources
    returned by lagoon-server"""

    INDEX_COL = "ix"
    INDEX_COL_TYPE = int

    def __init__(self, json):
        self.__json = json

    def __str__(self):
        return "<Source: ix={0}, view_name={1}>".format(self.ix, self.view_name)

    def __repr__(self):
        return str(self)

    @property
    def ix(self):
        return self.__json["ix"]

    @property
    def view_name(self):
        return self.__json["viewName"]

    @property
    def columns(self):
        return {get_column_name(c): c for c in self.__json["columns"]}

    @property
    def _json(self):
        return self.__json

    @property
    def schema(self):
        return self.__json["schema"]

    @property
    def col_types(self):
        types = {name: get_column_type(col) for name, col in self.columns.items()}
        types[self.INDEX_COL] = self.INDEX_COL_TYPE
        return types


def get_column_type(column: dict):
    column_type = column["type"]
    if column_type in LAGOON_TYPES:
        return LAGOON_TYPES[column_type]
    else:
        return UNKNOWN_COLUMN_TYPE


def get_column_name(column: dict):
    return column["inView"]
